tutorial/tests/test_object_oriented_programming.py

```diff
@@ -2,6 +2,221 @@
 
 import pytest
 
+#
+# Example 1: Person
+#
+
+
+def reference_oop_person(first_name: str, last_name: str):
+    class Person:
+        """A class representing a person with first name and last name"""
+
+        def __init__(self, first_name: str, last_name: str):
+            self.first_name = first_name
+            self.last_name = last_name
+
+    return Person(first_name, last_name)
+
+
+def validate_oop_person(solution_result):
+    params = list(vars(solution_result))
+    assert (
+        type(solution_result).__name__ == "Person"
+    ), "The class should be named 'Person'."
+    assert len(params) == 2, "The class should have 2 attributes."
+    assert (
+        "first_name" in params and "last_name" in params
+    ), "The class attributes should be 'first_name' and 'last_name'."
+
+
+@pytest.mark.parametrize(
+    "first_name, last_name",
+    [
+        ("Jane", "Doe"),
+        ("John", "Doe"),
+    ],
+)
+def test_oop_person(first_name, last_name, function_to_test):
+    solution_result = function_to_test(first_name, last_name)
+    reference_result = reference_oop_person(first_name, last_name)
+
+    validate_oop_person(solution_result)
+    assert (
+        solution_result.first_name == reference_result.first_name
+        and solution_result.last_name == reference_result.last_name
+    )
+
+
+#
+# Example 2: Person's full name
+#
+
+
+def reference_oop_fullname(first_name: str, last_name: str):
+    class Person:
+        """A class representing a person with first name and last name"""
+
+        def __init__(self, first_name: str, last_name: str):
+            self.first_name = first_name
+            self.last_name = last_name
+
+        def full_name(self) -> str:
+            return f"My name is {self.first_name} {self.last_name}"
+
+    return Person(first_name, last_name)
+
+
+def validate_oop_fullname(solution_result):
+    methods = [
+        attr
+        for attr in dir(solution_result)
+        if callable(getattr(solution_result, attr)) and not attr.startswith("__")
+    ]
+    assert len(methods) == 1, "The class should have 1 method."
+    assert "full_name" in methods, "The class method should be called 'full_name'."
+
+
+@pytest.mark.parametrize(
+    "first_name, last_name",
+    [
+        ("John", "Doe"),
+    ],
+)
+def test_oop_fullname(first_name, last_name, function_to_test):
+    solution_result = function_to_test(first_name, last_name)
+    reference_result = reference_oop_fullname(first_name, last_name)
+
+    validate_oop_person(solution_result)
+    validate_oop_fullname(solution_result)
+    assert (
+        solution_result.full_name() == reference_result.full_name()
+    ), "The full_name() result does not match the template 'My name is {first_name} {last_name}'."
+
+    solution_result.first_name = "Jane"
+    solution_result.last_name = "Smith"
+    reference_result.first_name = "Jane"
+    reference_result.last_name = "Smith"

```

Comment:
I don't quite get the purpose of these lines. Is this to "verify" that the return value is indeed a class with that attributes?

Reply:
If yes, then I don't think it's a good idea to "misuse" a class mutability.

Reply:
It is indeed a "misuse" of class mutability, but the purpose is to enforce the use of internal class attributes when writing the class method. It also took me a while to figure it out (see https://github.com/empa-scientific-it/python-tutorial/pull/249#discussion_r1836107872 above) but the problem is as follows:
- function `reference_oop_fullname()` has attributes `first_name` and `last_name`
- class `Person` also has attributes `first_name` and `last_name`
- when asking users to return the string `My name is {first_name} {last_name}`, if someone uses the function attributes directly, instead of the class attributes, then the tests will still pass
- by adding these lines in the code we make sure to test that they are using the `self.` attributes

Reply:
This is actually very interesting, because the "wrong" class which uses the solution function's parameters would be a **closure**. They would be using something straight from the advanced tutorial! 😄

Reply:
You could actually add a check like that instead of creating the "extra" person:

```python
def validate_oop_fullname(solution_result):
    methods = [
        attr
        for attr in dir(solution_result)
        if callable(getattr(solution_result, attr)) and not attr.startswith("__")
    ]
    assert len(methods) == 1, "The class should have 1 method."
    assert "full_name" in methods, "The class method should be called 'full_name'."
    assert solution_result.full_name.__closure__ is None, "full_name method should use class attributes (self.)"
```

And then get rid of the `different_person` line. Very elegant!

Reply:
This is awesome, I didn't think of it. Thanks! 🙌 Used it here: 1a058da395e3f17d329edf3efdab9f598a8aee1c